### PR TITLE
Update pow function for improved efficiency and handling of negative powers

### PR DESCRIPTION
**Pull request by mohammedouahman**
Update pow function for improved efficiency and handling of negative powers

This pull request updates the pow function in the Quaternion class to improve its efficiency and handle negative powers more effectively. The modifications include optimized algorithms and proper handling of negative powers using the inverse function. These changes enhance the overall performance and functionality of the Quaternion class.

* functions
  * Updated pow function in the Quaternion class for improved efficiency and handling of negative powers.

### CODE REVIEW

File: sympy/algebras/quaternion.py [hunk 1]
```diff
@@ -925,24 +925,19 @@ def pow(self, p):
...
         """
-        p = sympify(p)
-        q = self
         if p == -1:
-            return q.inverse()
-        res = 1
-
-        if not p.is_Integer:
+            return self.inverse()
+        if not isinstance(p, int):
```
Comment (smichr) [inline sympy/algebras/quaternion.py:930]: This sentinel must go first since `1 == 1.0`.

File: sympy/algebras/quaternion.py [hunk 2]
```diff
@@ -925,24 +925,19 @@ def pow(self, p):
...
             return NotImplemented
-
         if p < 0:
-            q, p = q.inverse(), -p
+            return self.inverse().pow(-p)
```
Comment (haru-44) [inline sympy/algebras/quaternion.py:933]: I don't think it is necessary to call pow again.

File: sympy/algebras/quaternion.py [hunk 3]
```diff
@@ -925,24 +926,20 @@ def pow(self, p):
...
         """
-        p = sympify(p)
-        q = self
-        if p == -1:
-            return q.inverse()
-        res = 1
-
-        if not p.is_Integer:
-            return NotImplemented
+        q, p = self, as_int(p)
```
Comment (oscarbenjamin) [inline sympy/algebras/quaternion.py:929]: The tests failed because `p` is not always an integer.

Comment (smichr) [inline sympy/algebras/quaternion.py:929]:
```suggestion
        try:
            q, p = self, as_int(p)
        except ValueError:
            return NotImplemented
```

**Pull request merged**